sys/drv/sd.c

```diff
@@ -0,0 +1,371 @@
+  int r;
+
+  uint64_t sup_blkcnt = 0;
+
+  if (state->props & SD_SUPP_CCS) {
+    /* Multiple block transfers either be terminated after a set amount of
+     * block is transferred, or by sending CMD_STOP_TRANS command */
+    if (num > 1 && (state->props & SD_SUPP_BLKCNT)) {
+      r = emmc_send_cmd(dev, EMMC_CMD(SET_BLOCK_COUNT), num, NULL);
+      if (r)
+        return 0;
+    }
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+    if (num > 1) {
+      emmc_send_cmd(dev, EMMC_CMD(READ_MULTIPLE_BLOCKS), lba, NULL);
+    } else {
+      emmc_send_cmd(dev, EMMC_CMD(READ_BLOCK), lba, NULL);
+    }
+    if (emmc_wait(dev, EMMC_I_READ_READY))
+      return 0;
+  } else {
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+  }
+  if (state->props & SD_SUPP_CCS) {
+    r = emmc_read(dev, buf, num * DEFAULT_BLKSIZE, NULL);
+    if (r)
+      return 0;
+    if (emmc_wait(dev, EMMC_I_DATA_DONE))
+      return 0;
+  } else {
+    for (uint32_t c = 0; c < num; c++) {
+      emmc_send_cmd(dev, EMMC_CMD(READ_BLOCK), lba + c, NULL);
+      r = emmc_read(dev, buf, 512, NULL);
+      if (r)
+        return 0;
+      if (emmc_wait(dev, EMMC_I_DATA_DONE))
+        return 0;
+      buf += 128;
+    }
+  }
+  if (num > 1 && !sup_blkcnt && (state->props & SD_SUPP_CCS))
+    emmc_send_cmd(dev, EMMC_CMD(STOP_TRANSMISSION), 0, NULL);
+  return num * DEFAULT_BLKSIZE;
+}
+
+/**
+ * Write a block to the sd card
+ * returns 0 on error
+ */
+int sd_write_blk(device_t *dev, uint32_t lba, void *buffer, uint32_t num) {
+  assert(dev->driver == (driver_t *)&sd_block_device_driver);
+  sd_state_t *state = (sd_state_t *)dev->state;
+
+  uint32_t *buf = (uint32_t *)buffer;
+  int r;
+  uint32_t c = 0;
+
+  if (num < 1)
+    return 0;
+
+  if (state->props & SD_SUPP_BLKCNT) {
+    if (num > 1 && (state->props & SD_SUPP_BLKCNT)) {
+      r = emmc_send_cmd(dev, EMMC_CMD(SET_BLOCK_COUNT), num, NULL);
+      if (r)
+        return 0;
+    }
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+
+    emmc_send_cmd(
+      dev, num == 1 ? EMMC_CMD(WRITE_BLOCK) : EMMC_CMD(WRITE_MULTIPLE_BLOCKS),
+      lba, NULL);
+    if (emmc_wait(dev, EMMC_I_WRITE_READY))
+      return 0;
+  } else {
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKCNT, num);
+    emmc_set_prop(dev, EMMC_PROP_RW_BLKSIZE, DEFAULT_BLKSIZE);
+  }
+  while (c < num) {
+    if (!(state->props & SD_SUPP_BLKCNT)) {
+      emmc_send_cmd(dev, EMMC_CMD(WRITE_BLOCK), lba + c, NULL);
+      if (emmc_wait(dev, EMMC_I_WRITE_READY))
+        return 0;
+    }
+    if (emmc_write(dev, buf, DEFAULT_BLKSIZE, NULL)) {
+      klog("e.MMC write failed");
+      return 0;
+    }
+    if (emmc_wait(dev, EMMC_I_DATA_DONE))
+      return 0;
+
+    c++;
+    buf += 128;
+  }
+
+  return num * DEFAULT_BLKSIZE;
+}
+
+static inline int sd_check_uio(uio_t *uio) {
```

Comment:
This function imposes a somehow strange constraint. Why can't I write 513 bytes to the SD card? In USB mass storage driver, in cases like this, I read the context of the block I'm about to overwrite only partially, and I write the unchanged part as it is. Why don't you  do that?

Reply:
Ok. Now I understand that it is the buffering layer which should take care of that.